feat: emv annotations graphics and text

deeonwuli opened the pull request:
### :pushpin: References

-   **Issue:** 
 Closes 
- [x] https://app.clickup.com/t/2ghcnmk
- [x] https://app.clickup.com/t/2ghcnn4
- [x] https://app.clickup.com/t/2ghcnmu


### :art: Screenshots

<img width="411" alt="Screenshot 2022-07-07 at 18 53 19" src="https://user-images.githubusercontent.com/37223065/177838131-3af2a6e4-3a7b-43d2-a512-793f43d63a42.png">
<img width="771" alt="Screenshot 2022-07-07 at 18 53 53" src="https://user-images.githubusercontent.com/37223065/177838210-9d6ea517-58d1-4242-bc29-b676e1bb9c18.png">

deeonwuli committed:
feat: emv model quality annotations graohic

deeonwuli committed:
Update ProtvistaPdbValidation.tsx

deeonwuli committed:
feat: basic forms of both charts

deeonwuli committed:
make cells clickable

tokland committed:
Add debugFlags to ease development

tokland commented on app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ProtvistaPdbValidation.tsx:
```diff
@@ -1,29 +1,204 @@
+                                ? "lime"
+                                : cellXValue === "2" && cellYValue === "3"
+                                ? "greenyellow"
+                                : "yellow",
```
Long object value. Create a local variable / separate function.

deeonwuli replied:
i'm adding the colours to the JSON file so i won't be using this anymore.

tokland reviewed (commented):
I've removed the example text and created some debug flags so you can work with a simplified view (which speeds up development) without breaking up the app. To use it, an example:

```
$ REACT_APP_HIDE_TRAINING=yes REACT_APP_SHOW_ONLY_VALIDATIONS=yes yarn start
```

I think Adrian has requested some visual changes, so I'll focus only on the code. Comments in-line. Let me know if you need clarification:

deeonwuli commented on app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ProtvistaPdbValidation.tsx:
```diff
@@ -1,29 +1,204 @@
+                data.push([]);
+                for (let column = 0; column < numCols; column++) {
+                    const cellXValue = modelQualityStats.data.categories[i]?.cat_id.split(",")[0];
+                    const cellYValue = modelQualityStats.data.categories[i]?.cat_id.split(",")[1];
```
i won't be making use of this anymore either.

deeonwuli committed:
Merge branch 'feature/emv-model-quality-annotations-graph' of https://github.com/EyeSeeTea/3DBIONOTES into feature/emv-model-quality-annotations-graph

deeonwuli committed:
make some pr corrections

deeonwuli commented on app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ProtvistaPdbValidation.tsx:
```diff
@@ -1,29 +1,204 @@
+            let height = 40;
+            let xPos = 1;
+            let yPos = height * numRows;
+            for (let row = 0; row < numRows; row++) {
```
i'm still trying to figure this out. i hope i get it done before i finish this task. but moving forward, i'll be using the functional style instead.

deeonwuli committed:
change text area for EM validation annotations

deeonwuli committed:
tidy up emv resolution graphic

deeonwuli committed:
add text to emv model local resolution stats

deeonwuli committed:
give rect no fill

deeonwuli committed:
fix tick placement

deeonwuli committed:
centralise text in grid cells

tokland commented on app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ProtvistaPdbValidation.tsx:
```diff
@@ -182,22 +198,190 @@ function useGrid() {
+    return svgRef;
+}
+
+interface gridData {
```
The convention naming for types is `CamelCase`.

tokland commented on app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ProtvistaPdbValidation.tsx:
```diff
@@ -36,141 +39,154 @@ function useGrid() {
                 data.push([]);
                 for (let column = 0; column < numCols; column++) {
-                    const cellXValue = modelQualityStats.data.categories[i]?.cat_id.split(",")[0];
-                    const cellYValue = modelQualityStats.data.categories[i]?.cat_id.split(",")[1];
```
Move reused code to local variable.

tokland commented on app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ProtvistaPdbValidation.tsx:
```diff
@@ -182,22 +198,190 @@ function useGrid() {
+
+        tooltip25
+            .append("text")
+            .text("Quartile-25: " + localResolutionStats.data.metrics[1]?.["quartile 25"] + " Å")
```
Not sure about this, just in case: I see that these two quartiles are shown on hover. But on the template of the ticket, it seems to always show all three values:

![image](https://user-images.githubusercontent.com/24643/177973390-5cf22102-e2c1-4f88-a9e7-6bc18c480e0d.png)

deeonwuli replied:
this was a change @adrianq requested. i was asked to ignore the median for now until the client confirms and work with just the 25th and 75th quartile values on hover.

tokland commented on app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ProtvistaPdbValidation.tsx:
```diff
@@ -36,141 +39,154 @@ function useGrid() {
             .style("text-anchor", "middle")
-            .attr("font-size", "0.35em")
+            .attr("font-size", "0.7em")
             .text("right");
```
`mapQ` literal is shown vertical, but it's horizontal on the screenshot.

tokland commented on app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ProtvistaPdbValidation.tsx:
```diff
@@ -182,22 +198,190 @@ function useGrid() {
+
+        tooltip75
+            .append("text")
+            .text("Quartile-75: " + localResolutionStats.data.metrics[2]?.["quartile 75"] + " Å")
```
Accesses looking like array[0], array[1], and so on are discouraged. Better unpack them so they have names and no explicit indexes are used:

```const [name, _ageWithUnderscoreIfUnused, addr = "Default value"] = userFieldsArray```

tokland reviewed (commented):
Looking much better! Some comments:

deeonwuli commented on app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ProtvistaPdbValidation.tsx:
```diff
@@ -2,29 +2,32 @@ import React from "react";
 import * as d3Module from "d3";
 import { ProtvistaPdb, ProtvistaPdbProps } from "./ProtvistaPdb";
 import modelQualityStats from "../../../data/repositories/emv_modelquality_stats.json";
+import localResolutionStats from "../../../data/repositories/emv_localresolution_stats.json";
```
i have this setting and the extension. i'll look into why it doesn't work.

deeonwuli committed:
pr fixes

deeonwuli committed:
increase grid height

tokland committed:
Enable eslint rule no-base-to-string

deeonwuli commented on app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ProtvistaPdbValidation.tsx:
```diff
@@ -69,71 +70,81 @@ function useGrid() {
             .attr("font-size", "0.75em")
             .style("text-anchor", "middle")
-            .text("0.45")
+            .text("0.45");
```
oh i forgot to switch out the hardcoded values for the values from the json. thank you so much for pointing this out.

deeonwuli commented on app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ProtvistaPdbValidation.tsx:
```diff
@@ -39,13 +39,14 @@ function useGrid() {
-            const data: gridData[][] = [];
+            const data: GridData[][] = [];
             const width = 80;
             const height = 60;
```
yeah, i discussed this with Adrian. he said same heights are fine for now until he gets some clarification from the client.

deeonwuli committed:
get tick values from json file

deeonwuli committed:
feat: add resolution median value

tokland commented on app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ProtvistaPdbValidation.tsx:
```diff
@@ -332,13 +359,48 @@ function useBar() {
         tooltip25
             .append("text")
-            .text("Quartile-25: " + quartile25 + " Å")
+            .text(`Quartile-25: ${Number(metrics["quartile 25"])}  Å`)
```
That will shown NaN if values is undefined. Maybe skip the whole tooltip25 if the metric is not found? Same for median and q75.

deeonwuli replied:
do you mean something like this? `${metrics["quartile 25"] ? Number(metrics["quartile 25"]) : ''}`

tokland replied:
I meant to completely skip the tooltip if we had nothing to show for it, but let's go with this and let Adrian decide. Only, it can be simplified: `.text(`Quartile-75: ${metrics["quartile 75"] ?? "-"}  Å`)`

deeonwuli replied:
okay got it.

deeonwuli committed:
Update ProtvistaPdbValidation.tsx

deeonwuli committed:
simplify tooltip text

tokland committed:
Prettify ProtvistaPdbValidation.tsx

tokland reviewed (approved):
Code looks good!

adrianq reviewed (changes requested):
@deeonwuli very good job here! A few comments, mainly styling and usability:
- When the mouse is over a cell in the matrix, can we make it a hand so it is clear for the user the cell is clickable?
- Numbers in the cell do not seem to be clickable
- Can we have the matrix and the local resolution bar on the same row? The text on the right of bar can go just below the bar.
- In the matrix, 'mapQ' and 'FSC-Q' can be bold and slightly bigger. They are the labels of axes and it should be clear for the user.
- It looks to me like the text close to the local resolution bar has a different font than the rest of the page, right?
- On hover on the percentiles in the bar, can we show the hand cursor?
- On hover in the bar, can we remove the two lines between the rectangle and the percentile? Just show the rectangle on hover close to the mouse. I am aware this was the original design but I don't it looks nice.
- In the bar, we will have only two vertical bars: per-25 and per-75. We will not have per-50 vertical bar.
- When hovering over the bars we will show a rectangle (same content for the two bards= containing three rows:
   - Median
   - Quartile-25
   - Quartile-75
- Very minor: it looks to me there is a double line in the axes (the cell + axis). Can we remove one of them?

deeonwuli committed:
make requested changes

- add pointer cursor on hover and click
- change layout of graphics
- font style fixes

deeonwuli committed:
feat: make adjustments to emv localResolutions graphic

adrianq commented:
@deeonwuli Thanks for the superfast modification! A few more comments:
- The matrix should use 50% of the width and the bar should use the other 50%. 
- If we increase the matrix width we can also increase the font of the labels in the axis
- Increase the font size of the numbers in the cell. 
- There should be no space between the bar and the text below. The tooltip can be displayed on top of the text below (superimpose)
- For the tooltip, can you mimic the styles of the tooltip that we have in the '?' close to the 'Basic Information' title?
![image](https://user-images.githubusercontent.com/6850223/181465822-1624a5da-749f-4787-8065-1152bcca06c9.png)

deeonwuli committed:
feat: update emv model quality and localRes graphics
